src/test/java/uk/nhs/hee/tis/trainee/forms/api/RestResponseEntityExceptionHandlerTest.java

```diff
@@ -0,0 +1,221 @@
+import org.springframework.context.support.DefaultMessageSourceResolvable;
+import org.springframework.core.MethodParameter;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.ProblemDetail;
+import org.springframework.http.ResponseEntity;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.validation.BeanPropertyBindingResult;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.method.MethodValidationResult;
+import org.springframework.validation.method.ParameterValidationResult;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.context.request.ServletWebRequest;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.method.annotation.HandlerMethodValidationException;
+import uk.nhs.hee.tis.trainee.forms.api.RestResponseEntityExceptionHandler.BodyValidationError;
+import uk.nhs.hee.tis.trainee.forms.api.RestResponseEntityExceptionHandler.ParameterValidationError;
+
+class RestResponseEntityExceptionHandlerTest {
+
+  private static final HttpStatusCode STATUS_CODE = HttpStatusCode.valueOf(
+      new Random().nextInt(400, 500));
+
+  private RestResponseEntityExceptionHandler handler;
+
+  @BeforeEach
+  void setUp() {
+    handler = new RestResponseEntityExceptionHandler();
+  }
+
+  @Test
+  void shouldHandleMethodArgumentNotValid() {
+    BeanPropertyBindingResult result = new BeanPropertyBindingResult(null, "dto");
+    result.addError(new FieldError("dto", "field2", "detail3"));
+    result.addError(new FieldError("dto", "field1", "detail2"));
+    result.addError(new FieldError("dto", "field1", "detail1"));
+    result.addError(new FieldError("dto", "field1.subField", "detail4"));
+    MethodArgumentNotValidException exception = new MethodArgumentNotValidException(
+        mock(MethodParameter.class), result);
+
+    HttpHeaders headers = HttpHeaders.EMPTY;
+    WebRequest request = new ServletWebRequest(new MockHttpServletRequest());
+
+    ResponseEntity<Object> response = handler.handleMethodArgumentNotValid(exception, headers,
+        STATUS_CODE, request);
+
+    assertThat("Unexpected response.", response, notNullValue());
+    assertThat("Unexpected headers.", response.getHeaders(), is(headers));
+    assertThat("Unexpected response code.", response.getStatusCode(), is(STATUS_CODE));
+    assertThat("Unexpected response type.", response.getBody(), instanceOf(ProblemDetail.class));
+
+    ProblemDetail problem = (ProblemDetail) response.getBody();
+    assertThat("Unexpected problem.", problem, notNullValue());
+    assertThat("Unexpected problem title.", problem.getTitle(), is("Validation failure"));
+    assertThat("Unexpected problem status.", problem.getStatus(), is(STATUS_CODE.value()));
+    assertThat("Unexpected problem status.", problem.getInstance(), nullValue());
```

Comment:
Probably should amend the error messages here.

Reply:
Nevermind, could see for looking. Will fix now.